---
title: Trypanosome Gene Structure Analysis (T. cruzi 1.3)
author: Keith Hughitt
output:
  html_document:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
  pdf_document:
    header-includes:
    - \usepackage{texshade}
---

Introduction
============

In previous work, *L. major* Friedlin and *T. cruzi* Y strain RNA-Seq data was
used to determine UTR boundaries and splice leader and polyadenylation acceptor
sites ([1](https://github.com/elsayed-lab/utr_analysis)), and to look for
evidence of alternative trans-splicing and poly-adenylation, and compute basic
statistics relating to SL and Poly(A) site usage
([2](http://www.umiacs.umd.edu/~keith/research/2015/110-utr-lengths/)) across
on a genome-wide scale.

Here, we expand on these analyses and attempt to map out the gene structure
for each of the parasites, and explore some of the basic properties and features 
of the trypanosome gene structure.

In order to ensure that the results are as easily reproduced as possible, all
relevant figures and tables are shown with the code used to generate them.

In a few instances where it is easier to extract the required information
outside of the R environment (for example, read mapping statistics), external
scripts are provided in the `scripts/` folder.

```{r include=FALSE}
library(knitr)
knit('settings/tcruzi_ystrain.Rmd', quiet=TRUE, output=tempfile())
```

```{r knitr_settings, include=FALSE}
opts_knit$set(progress=FALSE, verbose=TRUE)
opts_chunk$set(fig.width=1080/96,
               fig.height=1080/96,
               dpi=96,
               cache.path=CONFIG$cache_dir,
			   echo=CONFIG$verbose)
options(digits=2)
options(stringsAsFactors=FALSE)
options(knitr.duplicate.label='allow')

# If rmarkdown.pandoc.to not specified, have it default to 'latex' output
if (is.null(opts_knit$get("rmarkdown.pandoc.to"))) {
    opts_knit$set(rmarkdown.pandoc.to='latex')
}
```

```{r load_libraries, message=FALSE, warning=FALSE}
library('GenomicRanges')
library('Biostrings')
library('dynamicTreeCut')
library('rtracklayer')
library('edgeR')
library('ggplot2')
library('GO.db')
library('goseq')
library('gplots')
library('gridExtra')
library('reshape2')
library('MASS')
library('dplyr')
library('readr')
library('viridis')
library('msa')
library('seqLogo')
library('tibble')
source('../../../2015/00-shared/R/annotations.R')
source('../../../2015/00-shared/R/enrichment_analysis.R')
source('../../../2015/00-shared/R/util.R')
source('child/shared/helper.R')

# fix namespace for dplyr functions
assign('select', dplyr::select, envir=.GlobalEnv)

# random seed
set.seed(1)
```

```{r include=FALSE}
knit('child/shared/load_data.Rmd', quiet=TRUE, output=tempfile())
```

Results
=======

Coverage Statistics
-------------------

```{r child='child/shared/coverage_summary_statistics.Rmd'}
```

```{r child='child/shared/coverage_utrs.Rmd'}
```

UTR lengths
-----------

```{r child='child/shared/length_summary.Rmd'}
```

```{r child='child/shared/gene_structure_overview.Rmd'}
```

```{r child='child/shared/length_correlations.Rmd'}
```

Intergenic Regions
------------------

```{r child='child/shared/intergenic_regions.Rmd'}
```

Alternative trans-splicing and polyadenylation
----------------------------------------------

```{r child='child/shared/alternative_site_usage.Rmd'}
```

```{r child='child/shared/stage_specific_site_usage.Rmd'}
```

Polypyrimidine tracts
---------------------

```{r child='child/shared/polypyrimidine_tract.Rmd'}
```

Expression across developmental stages
--------------------------------------

```{r child='child/shared/expression_summary.Rmd'}
```

Mapping statistics
------------------

```{r child='child/shared/mapping_statistics.Rmd'}
```

System Info
===========

```{r}
sessionInfo()
```

```{r include=FALSE}
# write tables to file
table_dir <- file.path(CONFIG$output_dir, 'table')
dir.create(table_dir, recursive=TRUE)

for (table_name in ls()[grepl('^table', ls())]) {
    if (table_name %in% c('table_dir', 'table_name')) {
        next
    }
    filename <- sprintf("%s.csv", table_name)

    # write table to file with 2 sig figs
    write.csv(format(get(table_name), digits=1), 
              file=file.path(table_dir, filename))
}

# copy SL and Poly(A) gff files
data_dir <- file.path(CONFIG$output_dir, 'data')
dir.create(data_dir, recursive=TRUE)

for (gff in CONFIG$site_gffs) {
    file.copy(gff, file.path(data_dir, basename(gff)))
}

# copy HTML and figures
system(sprintf('(sleep 30 && cp -r README.html %s) &', CONFIG$output_dir))
system(sprintf('cp -r README_files/figure-html %s/figure', CONFIG$output_dir))
```

