## Primary vs. alternative acceptor sites

```{r}
# create combined site data frame including strand information
sl_sites_combined <- merge(sl_sites[['combined']], gene_strands, by='gene')
```

```{r, results='asis'}
# grab a window (90nt upstream, 10nt downstream) surrounding the acceptor site
win_size_upstream = 90
win_size_downstream = 10

sl_sites_combined <- sl_sites_combined %>% 
    mutate(acc_start=ifelse(strand == '+', 
                            chr_coordinate - win_size_upstream - 1, 
                            chr_coordinate - win_size_downstream),
           acc_end=ifelse(strand=='+', 
                          chr_coordinate + win_size_downstream, 
                          chr_coordinate + win_size_upstream + 1)) %>%
    mutate(acceptor_site_win=substr(chr_seqs[chr], acc_start, acc_end))

# reverse negative strand sequences
sl_sites_combined[sl_sites_combined$strand == '-',]$acceptor_site_win <- as.character(reverseComplement(
    DNAStringSet(sl_sites_combined[sl_sites_combined$strand == '-',]$acceptor_site_win)
))

# actual splice acceptor site is 2nt wide
sl_sites_combined <- sl_sites_combined %>% 
    mutate(acceptor_site=substr(acceptor_site_win, win_size_upstream + 1, win_size_upstream + 2),
           acceptor_site_3nt=substr(acceptor_site_win, win_size_upstream, win_size_upstream + 2))

# exclude any sites containing N's (not very common)
sl_sites_combined <- sl_sites_combined %>% 
    filter(str_detect(acceptor_site, '[AGCT]{2}'))

# in cases where primary site was excluded, remove gene from consideration
primary_site_genes <- (sl_sites_combined %>% filter(type == 'primary'))$gene

sl_sites_combined <- sl_sites_combined[sl_sites_combined$gene %in%
                                       primary_site_genes,]
```

```{r table_s11, results='asis'}
# total number of sites and number of primary sites
num_total <- nrow(sl_sites_combined)
num_primary <- nrow(sl_sites_combined %>% filter(type == 'primary'))
num_alternative <- nrow(sl_sites_combined %>% filter(type == 'alternative'))

primary_site_props <- sl_sites_combined %>% 
    filter(type == 'primary') %>%
    group_by(acceptor_site) %>% 
    summarize(primary=round((n() / num_primary) * 100, 2))

alt_site_props <- sl_sites_combined %>% 
    filter(type == 'alternative') %>%
    group_by(acceptor_site) %>% 
    summarize(alternative=round((n() / num_alternative) * 100, 2))

table_s11 <- merge(primary_site_props, alt_site_props, by='acceptor_site') %>%
    arrange(desc(primary))
colnames(table_s11) <- c('Acceptor sequence', 'Primary site', 'Minor site')
kable(table_s11, caption='**Table S11** Primary and alternative acceptor site sequence composition')
```

```{r figure_03a, fig.cap="**Figure 03a** Sequence composition of primary and alternative SL acceptor sites"}
sl_summary <- sl_sites_combined %>% 
    select(acceptor_site, type) %>%
    group_by(type, acceptor_site) %>% 
    summarise (num=n()) %>%
    mutate(prop=num / sum(num)) %>%
    arrange(desc(prop))

ggplot(sl_summary, aes(acceptor_site, y=prop, fill=type)) +
    geom_bar(position="dodge", stat='identity') +
    xlab("Acceptor site sequence") + 
    ylab("Acceptor site motif proportion")
```

## Acceptor site sequence logos

### Primary acceptor site logo

```{r figure_03e, fig.width=3840/192, fig.height=720/192, dpi=192, fig.cap="**Figure 03e** Sequence logo for region surrounding primary trans-splicing acceptor site."}
primary_seqs <- (sl_sites_combined %>% filter(type=='primary'))$acceptor_site_win
names(primary_seqs) <- (sl_sites_combined %>% filter(type=='primary'))$gene

# for logo purposes, exclude sequences with N's (~1%)
primary_seqs <- primary_seqs[str_count(primary_seqs, 'N') == 0]

# convert to a PFM, and then to a PWM
pwm <- consensusMatrix(primary_seqs, as.prob=TRUE)[DNA_BASES,]

# draw sequence logo
seqLogo(pwm, ic.scale=FALSE)
```

#### Sub-groups with similar primary acceptor site sequences

```{r figure_XX_sl_acceptor_site_sequence_heatmap}
# measure pairwise hamming distances between all acceptor site sequences
num_seqs <- length(primary_seqs)
hamming_dist_mat <- as.matrix(stringdistmatrix(primary_seqs, method='hamming', useNames='names'))

# convert to similarity matrix
seq_len <- nchar(primary_seqs[1])
sim_mat <- 1 - (hamming_dist_mat / seq_len)

# set diagonal to zero (improves dynamic range of other values)
diag(sim_mat) <- 0

# for visualization purposes, we will we will subsample the sequences to speed
# things up a bit
ind <- sample(1:nrow(sim_mat), 1000)

heatmap.2(sim_mat[ind,ind], trace='none', 
          symkey=FALSE, cex_lab=0.3, col=viridis, labCol=FALSE,
          labRow=FALSE,
          main='Pairwise Trans-splicing Acceptor Site Region Sequence Alignments')
```

```{r acceptor_site_cluster_sequences, results='asis', cache=TRUE, autodep=TRUE}
# convert to distance matrix and find clusters of similar acceptor site area
# sequences 
dist_mat <- 1 - ((sim_mat - min(sim_mat)) / max(sim_mat))
tree <- hclust(as.dist(dist_mat), method='average')

# work-around for cutree bug
tree$height = round(tree$height, 10)

#clusters <- cutree(tree, h=0.6)
#plot(tree, xlab="Polypyrimidine tract sequences", sub="", labels=FALSE)
#abline(h=0.6, col='red', lty=2)

clusters <- cutreeDynamicTree(dendro=tree, minModuleSize=30)
#table(clusters)

# get average similarity within each cluster
avg_similarities <- c()

for (i in sort(unique(clusters))) {
    cluster_ind <- clusters == i

    avg_similarities <- c(avg_similarities, median(sim_mat[cluster_ind,
                                                           cluster_ind]))
}

df <- data_frame(cluster=sort(unique(clusters)), 
                 avg_similarity=avg_similarities,
                 num_seqs=table(clusters))
df <- df %>% arrange(desc(avg_similarity))
```

```{r figure_XX_sl_acceptor_site_sequence_cluster, results='asis', cache=TRUE, autodep=TRUE, fig.width=3840/192, fig.height=720/192, dpi=192}
# Iterate over top N clusters and generate a sequence logo for each
n <- 10

for (i in 1:n) {
    cluster_choice <- (df %>% arrange(desc(avg_similarity)))$cluster[i]
    cluster_seqs <- primary_seqs[clusters == cluster_choice]

    # convert to a PFM, and then to a PWM
    pwm <- consensusMatrix(cluster_seqs, as.prob=TRUE)[DNA_BASES,]

    # draw sequence logo
    cat(sprintf("\n##### Primary acceptor site cluster %d (%d genes)\n\n", i, df$num_seqs[i]))

    seqLogo(pwm, ic.scale=FALSE)

    cat('\n')
}
```

### Alternative acceptor site logo

```{r figure_s05, fig.width=3840/192, fig.height=720/192, dpi=192, fig.cap="**Figure S5** Sequence logo for region surrounding alternative trans-splicing acceptor sites."}

alt_seqs <- (sl_sites_combined %>% filter(type=='alternative'))$acceptor_site_win

# for logo purposes, exclude sequences with N's (~1%)
alt_seqs <- alt_seqs[str_count(alt_seqs, 'N') == 0]

# convert to a PFM, and then to a PWM
pwm <- consensusMatrix(alt_seqs, as.prob=TRUE)[DNA_BASES,]

# draw sequence logo
seqLogo(pwm, ic.scale=FALSE)
```

## Distance between primary and alternative trans-splicing sites

### Canonical AG acceptor site

```{r figure_03b, results='asis', warning=FALSE, fig.cap="**Figure 03b** Distribution of distances between primary and alternative trans-splicing acceptor sites for canonical acceptor sites."}
sl_sites_combined <- sl_sites_combined %>% 
    group_by(gene) %>% 
    mutate(primary_site_dist=ifelse(strand=='+', chr_coordinate - chr_coordinate[type == 'primary'],
                                                 chr_coordinate[type == 'primary'] - chr_coordinate)) %>%
    mutate(orientation=ifelse(primary_site_dist < 0, 'upstream', 'downstream'))

# primary - minor acceptor site distances
ag_dists <- (sl_sites_combined %>% 
             filter(type == 'alternative' & acceptor_site == 'AG'))$primary_site_dist

# convert to data frame and add color based on orientation
ag_dists <- data.frame(dist=ag_dists) %>%
    mutate(orientation=ifelse(dist < 0, 'upstream', 'downstream'))

ggplot(ag_dists, aes(x=dist, fill=orientation)) + 
    geom_histogram(binwidth=20, center=10, color='white') +
    xlim(-1000, 1000)
```

```{r}
ag_dists %>% 
    group_by(orientation) %>%
    summarize(number=n(), proportion=n() / nrow(ag_dists)) %>%
    kable(caption='**Table X** Average distance between primary and alternative acceptor sites for canonical trans-splicing sites')
```

### Non-canonical acceptor sites

```{r, figure_03c, warning=FALSE, fig.cap="**Figure 03c** Distribution of distances between primary and alternative trans-splicing acceptor sites for non-canonical acceptor sites."}
ag_dists <- (sl_sites_combined %>% 
             filter(type == 'alternative' & acceptor_site != 'AG'))$primary_site_dist

# convert to data frame and add color based on orientation
ag_dists <- data.frame(dist=ag_dists) %>%
    mutate(orientation=ifelse(dist < 0, 'upstream', 'downstream'))

ggplot(ag_dists, aes(x=dist, fill=orientation)) + 
    geom_histogram(binwidth=10, center=5, color='white') +
    xlim(-1000, 1000)
```

```{r}
ag_dists %>% 
    group_by(orientation) %>%
    summarize(number=n(), proportion=n() / nrow(ag_dists)) %>%
    kable(caption='**Table X** Average distance between primary and alternative acceptor sites for non-canonical trans-splicing sites')
```

## Preceeding nucleotide

```{r, table_s12}
# primary site proportions (including upstream nucleotide)
primary_site_props <- sl_sites_combined %>%
    filter(type == 'primary') %>%
    group_by(acceptor_site_3nt) %>%
    summarise (num=n()) %>%
    mutate(prop=num / sum(num)) %>%
    filter(acceptor_site_3nt %in% c('AAG', 'CAG', 'GAG', 'TAG')) %>%
    select(-num) %>%
    arrange(desc(prop))

# alt site proportions, divided by up- and down-stream orientation, relative
# to the primary site
alt_site_props <- sl_sites_combined %>%
    filter(type=='alternative') %>%
    group_by(orientation, acceptor_site_3nt) %>%
    summarise (num=n()) %>%
    mutate(primary=num / sum(num)) %>%
    filter(acceptor_site_3nt %in% c('AAG', 'CAG', 'GAG', 'TAG')) %>%
    select(-num) %>% 
    arrange(desc(primary)) %>%
    spread(orientation, primary)

table_s12 <- merge(primary_site_props, alt_site_props, by='acceptor_site_3nt')
colnames(table_s12) <- c('site', 'Primary', 'Alternative downstream', 
                         'Alternative upstream')
table_s12 <- table_s12 %>% column_to_rownames('site')
table_s12 <- round(table_s12 * 100)

kable(table_s12, caption='**Table S12** Frequency of preceeding nucleotide for primary and alternative trans-splicing acceptor sites.')
```

