## Primary vs. alternative acceptor site

```{r}
# create combined site data frame including strand information
sl_sites_combined <- merge(sl_sites[['combined']], gene_strands, by='gene')
```

```{r, results='asis'}
# grab a window (90nt upstream, 10nt downstream) surrounding the acceptor site
win_size_upstream = 90
win_size_downstream = 10

sl_sites_combined <- sl_sites_combined %>% 
    mutate(acc_start=ifelse(strand == '+', 
                            chr_coordinate - win_size_upstream - 1, 
                            chr_coordinate - win_size_downstream),
           acc_end=ifelse(strand=='+', 
                          chr_coordinate + win_size_downstream, 
                          chr_coordinate + win_size_upstream + 1)) %>%
    mutate(acceptor_site_win=substr(chr_seqs[chr], acc_start, acc_end))

# reverse negative strand sequences
sl_sites_combined[sl_sites_combined$strand == '-',]$acceptor_site_win <- as.character(reverseComplement(
    DNAStringSet(sl_sites_combined[sl_sites_combined$strand == '-',]$acceptor_site_win)
))

# actual splice acceptor site is 2nt wide
sl_sites_combined <- sl_sites_combined %>% 
    mutate(acceptor_site=substr(acceptor_site_win, win_size_upstream + 1, win_size_upstream + 2),
           acceptor_site_3nt=substr(acceptor_site_win, win_size_upstream, win_size_upstream + 2))

# exclude any sites containing N's (not very common)
sl_sites_combined <- sl_sites_combined %>% 
    filter(str_detect(acceptor_site, '[AGCT]{2}'))

# in cases where primary site was excluded, remove gene from consideration
primary_site_genes <- (sl_sites_combined %>% filter(type == 'primary'))$gene

sl_sites_combined <- sl_sites_combined[sl_sites_combined$gene %in%
                                       primary_site_genes,]
```

```{r table_S6, results='asis'}
# total number of sites and number of primary sites
num_total <- nrow(sl_sites_combined)
num_primary <- nrow(sl_sites_combined %>% filter(type == 'primary'))
num_alternative <- nrow(sl_sites_combined %>% filter(type == 'alternative'))

primary_site_props <- sl_sites_combined %>% 
    filter(type == 'primary') %>%
    group_by(acceptor_site) %>% 
    summarize(primary=round((n() / num_primary) * 100, 2))

alt_site_props <- sl_sites_combined %>% 
    filter(type == 'alternative') %>%
    group_by(acceptor_site) %>% 
    summarize(alternative=round((n() / num_alternative) * 100, 2))

table_S6 <- merge(primary_site_props, alt_site_props, by='acceptor_site') %>%
    arrange(desc(primary))
colnames(table_S6) <- c('Acceptor sequence', 'Primary site', 'Minor site')
kable(table_S6, caption='**Table S6** Primary and alternative acceptor site sequence composition')
```

```{r figure_2a_sl_sequence_composition, fig.cap="**Figure 2a** Sequence composition of primary and alternative SL acceptor sites"}
sl_summary <- sl_sites_combined %>% 
    select(acceptor_site, type) %>%
    group_by(type, acceptor_site) %>% 
    summarise (num=n()) %>%
    mutate(prop=num / sum(num)) %>%
    arrange(desc(prop))

# manually specify order in which primary/alternatie bars appear
sl_summary$type <- factor(sl_summary$type, levels=c('primary', 'alternative'))

ggplot(sl_summary, aes(reorder(acceptor_site, -num), y=prop, fill=type)) +
    geom_bar(position="dodge", stat='identity') +
    xlab("Acceptor site sequence") + 
    ylab("Acceptor site motif proportion")
```

## Acceptor site sequence logos

### Primary acceptor site logo

```{r figure_2e_trans_splicing_site_seqlogo, warning=FALSE, fig.width=3840/192, fig.height=720/192, dpi=192, fig.cap="**Figure 2e** Sequence logo for region surrounding primary trans-splicing acceptor site."}
primary_seqs <- (sl_sites_combined %>% filter(type=='primary'))$acceptor_site_win
names(primary_seqs) <- (sl_sites_combined %>% filter(type=='primary'))$gene

# for logo purposes, exclude sequences with N's (~1%)
primary_seqs <- primary_seqs[str_count(primary_seqs, 'N') == 0]

# convert to a PFM, and then to a PWM
pwm <- consensusMatrix(primary_seqs, as.prob=TRUE)[DNA_BASES,]

# draw sequence logo
seqLogoMod(pwm, xaxis=-91:10, ic.scale=TRUE)
```

#### Sub-groups with similar primary acceptor site sequences

Next, in order to look for possible groups of genes with similar sequences for
the trans-splicing acceptor site and surrounding regions, a heatmap of the
pairwise sequence similarities for the same region depicted in the above
sequence logo is generated.

Each row and column represents the 100nt acceptor site region sequence for a
single gene, with blocks along the diagonal indicating clusters of similar
sequences.

```{r figure_X5_sl_acceptor_site_sequence_heatmap, warning=FALSE, fig.cap='**Figure X5** Heatmap of primary site acceptor sequence similarity'}
# measure pairwise hamming distances between all acceptor site sequences
num_seqs <- length(primary_seqs)
hamming_dist_mat <- as.matrix(stringdistmatrix(primary_seqs, method='hamming', useNames='names'))

# convert to similarity matrix
seq_len <- nchar(primary_seqs[1])
sim_mat <- 1 - (hamming_dist_mat / seq_len)

# set diagonal to zero (improves dynamic range of other values)
diag(sim_mat) <- 0

# for visualization purposes, we will we will subsample the sequences to speed
# things up a bit
ind <- sample(1:nrow(sim_mat), 1000)

heatmap.2(sim_mat[ind,ind], trace='none', 
          symkey=FALSE, cex_lab=0.3, col=viridis, labCol=FALSE,
          labRow=FALSE,
          main='Pairwise TS Acceptor Site Region Sequence Alignments')
```

```{r acceptor_site_cluster_sequences, results='asis', cache=TRUE, autodep=TRUE}
# convert to distance matrix and find clusters of similar acceptor site area
# sequences 
dist_mat <- 1 - ((sim_mat - min(sim_mat)) / max(sim_mat))
tree <- hclust(as.dist(dist_mat), method='average')

# work-around for cutree bug
tree$height = round(tree$height, 10)

#clusters <- cutree(tree, h=0.6)
#plot(tree, xlab="Polypyrimidine tract sequences", sub="", labels=FALSE)
#abline(h=0.6, col='red', lty=2)

clusters <- cutreeDynamicTree(dendro=tree, minModuleSize=30)
#table(clusters)

# get average similarity within each cluster
avg_similarities <- c()

for (i in sort(unique(clusters))) {
    cluster_ind <- clusters == i

    avg_similarities <- c(avg_similarities, median(sim_mat[cluster_ind,
                                                           cluster_ind]))
}

df <- data_frame(cluster=sort(unique(clusters)), 
                 avg_similarity=avg_similarities,
                 num_seqs=table(clusters))
df <- df %>% arrange(desc(avg_similarity))
```

To see what the acceptor site region motif looks like for clusters of genes
with similar sequences in this region, sequence logos are shown for the 10
clusters of genes with the highest intra-cluster sequence similarity.

```{r figure_X6_sl_acceptor_site_sequence_cluster, results='asis', cache=TRUE, autodep=TRUE, fig.width=3840/192, fig.height=720/192, dpi=192, fig.cap='**Figure X6** Primary acceptor site sequence motif clusters.'}
# Iterate over top N clusters and generate a sequence logo for each
n <- 10

for (i in 1:n) {
    cluster_choice <- (df %>% arrange(desc(avg_similarity)))$cluster[i]
    cluster_seqs <- primary_seqs[clusters == cluster_choice]

    # convert to a PFM, and then to a PWM
    pwm <- consensusMatrix(cluster_seqs, as.prob=TRUE)[DNA_BASES,]

    # draw sequence logo
    cat(sprintf("\n##### Primary acceptor site cluster %d (%d genes)\n\n", i, df$num_seqs[i]))
    
    # update caption and plot
    opts_chunk$set('fig.cap', sprintf('**Figure X6 (%d/10)** Primary acceptor site sequence motif clusters.', i)) 
    seqLogoMod(pwm, xaxis=-91:10, ic.scale=TRUE)

    cat('\n\n\n')
}

opts_chunk$set('fig.cap', NULL)
```

### Alternative acceptor site logo

```{r figure_S3, fig.width=3840/192, fig.height=720/192, dpi=192, fig.cap="**Figure S3** Sequence logo for region surrounding alternative trans-splicing acceptor sites."}

alt_seqs <- (sl_sites_combined %>% filter(type=='alternative'))$acceptor_site_win

# for logo purposes, exclude sequences with N's (~1%)
alt_seqs <- alt_seqs[str_count(alt_seqs, 'N') == 0]

# convert to a PFM, and then to a PWM
pwm <- consensusMatrix(alt_seqs, as.prob=TRUE)[DNA_BASES,]

# draw sequence logo
seqLogoMod(pwm, xaxis=-91:10, ic.scale=TRUE)
```

## Distance between primary and alternative trans-splicing sites

### Canonical AG acceptor site

```{r figure_2b_canonical_trans_splice_site_primary_minor_dists, results='asis', warning=FALSE, fig.cap="**Figure 2b** Distribution of distances between primary and alternative trans-splicing acceptor sites for canonical acceptor sites."}
sl_sites_combined <- sl_sites_combined %>% 
    group_by(gene) %>% 
    mutate(primary_site_dist=ifelse(strand=='+', chr_coordinate - chr_coordinate[type == 'primary'],
                                                 chr_coordinate[type == 'primary'] - chr_coordinate)) %>%
    mutate(orientation=ifelse(primary_site_dist < 0, 'upstream', 'downstream'))

canonical_sl_sites <- sl_sites_combined %>% 
    filter(acceptor_site == 'AG')

# primary - minor acceptor site distances
ag_dists <- (canonical_sl_sites %>% 
             filter(type == 'alternative' & acceptor_site == 'AG'))$primary_site_dist

# convert to data frame and add color based on orientation
ag_dists <- data.frame(dist=ag_dists) %>%
    mutate(orientation=ifelse(dist < 0, 'upstream', 'downstream'))

ggplot(ag_dists, aes(x=dist, fill=orientation)) + 
    geom_histogram(binwidth=20, center=10, color='white') +
    xlim(-1000, 1000)
```

```{r, table_X7}
table_X7 <- ag_dists %>% 
    group_by(orientation) %>%
    summarize(number=n(), proportion=n() / nrow(ag_dists))
kable(table_X7, caption='**Table X7** Average distance between primary and alternative acceptor sites for canonical trans-splicing sites')
```

### Non-canonical acceptor sites

```{r, figure_2c_noncanonical_trans_splice_site_primary_minor_dists, warning=FALSE, fig.cap="**Figure 2c** Distribution of distances between primary and alternative trans-splicing acceptor sites for non-canonical acceptor sites."}
ag_dists <- (sl_sites_combined %>% 
             filter(type == 'alternative' & acceptor_site != 'AG'))$primary_site_dist

# convert to data frame and add color based on orientation
ag_dists <- data.frame(dist=ag_dists) %>%
    mutate(orientation=ifelse(dist < 0, 'upstream', 'downstream'))

ggplot(ag_dists, aes(x=dist, fill=orientation)) + 
    geom_histogram(binwidth=10, center=5, color='white') +
    xlim(-1000, 1000)
```

```{r, table_X8}
table_X8 <- ag_dists %>% 
    group_by(orientation) %>%
    summarize(number=n(), proportion=n() / nrow(ag_dists))
kable(table_X8, caption='**Table X8** Average distance between primary and alternative acceptor sites for non-canonical trans-splicing sites')
```

## Preceeding nucleotide for canonical SL sites

This table includes the percentage usage for each trinucleotide XAG acceptor
site motif: CAG, AAG, TAG, and GAG, separated by primary and up- or down-stream
alternative sites.

```{r, table_X9}
# primary site proportions (including upstream nucleotide)
primary_site_props <- canonical_sl_sites %>%
    filter(type == 'primary') %>%
    group_by(acceptor_site_3nt) %>%
    summarise (num=n()) %>%
    mutate(prop=num / sum(num)) %>%
    select(-num) %>%
    arrange(desc(prop))

# alt site proportions, divided by up- and down-stream orientation, relative
# to the primary site
alt_site_props <- canonical_sl_sites %>%
    filter(type=='alternative') %>%
    group_by(orientation, acceptor_site_3nt) %>%
    summarise (num=n()) %>%
    mutate(primary=num / sum(num)) %>%
    select(-num) %>% 
    arrange(desc(primary)) %>%
    spread(orientation, primary)

table_X9 <- merge(primary_site_props, alt_site_props, by='acceptor_site_3nt')
colnames(table_X9) <- c('site', 'Primary', 'Alternative downstream', 
                         'Alternative upstream')
table_X9 <- table_X9 %>% 
    arrange(desc(Primary)) %>%
    column_to_rownames('site')
table_X9 <- round(table_X9 * 100, 2)
table_X9[is.na(table_X9)] <- 0

kable(table_X9, caption='**Table X9** Frequency of preceeding nucleotide for primary and alternative canonical trans-splicing acceptor sites.')
```

