## RNA-Seq coverage support for UTR boundaries

```{r table_S1, results='asis'}
# coverage statistics collected from original input RNA-Seq fastq files,
# and UTR pipeline intermediate files.
# see also: scripts/*/S01_*   
num_reads$condition_long <- CONFIG$condition_mapping$to[match(num_reads$condition, 
                                                        CONFIG$condition_mapping$from)]

table_S1 <- num_reads %>% 
    mutate(pct_reads_sl=(num_reads_sl/num_reads_parasite) * 100,
           pct_reads_polya=(num_reads_polya/num_reads_parasite) * 100) %>%
    group_by(condition_long) %>%
    summarize("# reads mapped to host"=sum(num_reads_host, na.rm=TRUE),
              "# reads mapped to parasite"=sum(num_reads_parasite),
              "# SL-containing reads"=sum(num_reads_sl),
              "% SL-containing reads"=mean(pct_reads_sl),
              "# Poly(A)-containing reads"=sum(num_reads_polya),
              "% Poly(A)-containing reads"=mean(pct_reads_polya)) %>%
    rename("Parasite Stage"=condition_long)

xkable(table_S1, caption='**Table S1**: Summary of mapping statistics.')
```

```{r results='asis'}
cat(sprintf('- Total number of SL-containing reads: %d (%0.3f%%)\n', 
            sum(num_reads$num_reads_sl),
            sum(num_reads$num_reads_sl) / sum(num_reads$num_reads_parasite) * 100))
cat(sprintf('- Total number of Poly(A)-containing reads: %d (%0.3f%%)\n\n', 
            sum(num_reads$num_reads_polya),
            sum(num_reads$num_reads_polya) / sum(num_reads$num_reads_parasite) * 100))
```

```{r, results='asis', eval=FALSE}
#
# DEPRECATED
#
# Old table S1; dropped from manuscript Nov 28, 2016 since most of these stats
# have already been in a previous manuscript
#

# human-readable condition names
num_reads$condition_long <- CONFIG$condition_mapping$to[match(num_reads$condition, 
                                                        CONFIG$condition_mapping$from)]
table_S1 <- num_reads %>%
    mutate("Infected"=ifelse(condition %in% c('proc', 'meta', 'epi', 'trypo'), 'N', 'Y'),
           num_reads_host=replace(num_reads_host, is.na(num_reads_host), 0),
           num_reads_parasite=replace(num_reads_parasite, is.na(num_reads_parasite), 0)) %>%
    mutate(total_mapped=num_reads_host + num_reads_parasite) %>%
    mutate(total_mapped_pct=(total_mapped/num_reads_total) * 100,
           parasite_mapped_pct=(num_reads_parasite/total_mapped) * 100,
           host_mapped_pct=(num_reads_host/total_mapped) * 100) %>%
    select("Lab sample ID"=sample_id,
           "Accession"=accession,
           "Lab sample ID"=sample_id, 
           "Developmental Stage"=condition_long,
           Infected,
           "Batch"=batch,
           "Total number of input reads"=num_reads_total,
           "Total number of reads mapped"=total_mapped,
           "% of total reads mapped"=total_mapped_pct,
           "Reads mapped to parasite"=num_reads_parasite,
           "Reads mapped to hg19"=num_reads_host,
           "% of mapped reads belonging to parasite"=parasite_mapped_pct,
           "% of mapped reads belonging to hg19"=host_mapped_pct)
xkable(table_S1, digits=3, 
       caption='**Table S1**: Summary of samples collected and mapping statistics')
```

