## Average gene RPKM expression by chromosome

The next section represents a brief deviation from the focus on gene structure.
Here instead, the focus is on the global level of expression across parasite
chromosomes and polycistronic developmental units (PTUs), both globally and as
a function of developmental stage.

To begin, we explore the average expression level (in RPKM) as a function of 
chromosome. This is useful for indicating possible chromosomal duplications.

In the first plot, the average expression at the gene level is shown by
chromosome.

```{r data_prep}
# Create RPKM-normalized version of counts
gene_lengths <- data.frame(gene=genes$ID, length=as.numeric(genes$size))

# remove duplicates and entries missing from main count table
gene_lengths <- gene_lengths[gene_lengths$gene %in% rownames(count_table),]
gene_lengths <- gene_lengths[!duplicated(gene_lengths$gene),]

# Exclude counts for SLRNAs, etc. not included in gene_lengths
count_table <- count_table[rownames(count_table) %in% gene_lengths$gene,]

# Create a vector of gene lengths
length_vector <- gene_lengths[match(row.names(count_table), gene_lengths$gene), 'length']

rpkm_counts <- rpkm(count_table, gene.length=length_vector)
```

```{r figure_X11_average_chr_gene_rpkm_expr, fig.cap='**Figure X11** Average RPKM expression by chromosome at the gene level'}
# Get total expression for each gene, across all samples
rpkm_totals <- rownames_to_column(as.data.frame(rowSums(rpkm_counts)))
colnames(rpkm_totals) <- c('gene', 'expr')

# Add chromosome information
rpkm_totals <- merge(rpkm_totals, chr_mapping, by='gene') 

# Compute average total RPKM for each chromosome
chr_expr_averages <- rpkm_totals %>% 
    group_by(chr) %>%
    summarize(average_expr=median(expr),
              total_expr=sum(expr))

ggplot(chr_expr_averages, aes(x=chr, y=average_expr)) + 
    geom_bar(stat='identity', color=CONFIG$stroke, fill=CONFIG$colors[1]) +
    scale_x_discrete(limits=CONFIG$chr_order) +
    theme(axis.text.x=element_text(angle=90)) +
    xlab("Chromosome") +
    ylab("Median Gene Expression (RPKM)") +
    ggtitle("Average RPKM expression by chromosome")
```

## Average RPKM expression by chromosome at the nucleotide level

Next, the average expression per nucleotide is shown for each chromosome. This
is an alternate way to get at the "average" expression of genes on a given
chromosome.

```{r figure_X12_average_chr_rpkm_expr_by_nt, fig.cap='**Figure X12** Average RPKM expression by chromosome at the nucleotide level'}
chr_cds_total_lengths <- merge(gene_lengths, chr_mapping, by='gene') %>%
    group_by(chr) %>%
    summarise(total_length=sum(length))

chr_stats <- merge(chr_cds_total_lengths, chr_expr_averages, by='chr') %>%
    mutate(expr_per_nt=total_expr/total_length)

ggplot(chr_stats, aes(x=chr, y=expr_per_nt)) +
    geom_bar(stat='identity', color=CONFIG$stroke, fill=CONFIG$colors[1]) +
    scale_x_discrete(limits=CONFIG$chr_order) +
    theme(axis.text.x=element_text(angle=90)) +
    xlab("Chromosome") +
    ylab("Median RPKM expression per nucleotide") +
    ggtitle("Average RPKM expression per nt for each chromosome")
```

## Average RPKM expression by chromosome (extracellular vs. intracellular)

Below, the same plots are shown, but with expression divided among the
intracellular and extracellular parasite developmental stages. Any large
discrepancies here would indicate chromosome-level expression changes across
developmental stage.

```{r figure_X13_average_chr_gene_rpkm_expr_stage_specific, fig.cap='**Figure X13** Average RPKM expression by chromosome at the gene level (intra vs extracellular)'}
# Create a long version of RPKM counts
rpkm_counts_long <- melt(rpkm_counts)
colnames(rpkm_counts_long) <- c('gene', 'sample_id', 'expr')

# Add stage information
rpkm_counts_long <- merge(rpkm_counts_long, CONFIG$samples %>%
                          select(sample_id, condition, stage),
                          by='sample_id')

# Add chromosome information
rpkm_counts_long <- merge(rpkm_counts_long, chr_mapping, by='gene') 

# Compute average stage-specific median RPKM for each chromosome 
# (intracellular vs. extracellular)
chr_expr_averages <- rpkm_counts_long %>% 
    group_by(chr, stage) %>%
    summarize(average_expr=median(expr),
              total_expr=sum(expr))

ggplot(chr_expr_averages, aes(x=stage, y=average_expr, fill=stage)) + 
    geom_bar(stat='identity', color=CONFIG$stroke) +
    scale_fill_manual(values=CONFIG$colors) +
    theme(axis.text.x=element_text(angle=90)) +
    facet_wrap(~chr)
```

## Total RPKM expression by chromosome (extracellular vs. intracellular)

For comparison, a similar plot is shown but using the _total_ expression across
stage and chromosome, instead of the _average_ expression.

```{r figure_X14_total_chr_rpkm_expr_stage_specific, fig.cap='**Figure X14** Total RPKM expression by chromosome at the gene level (intra vs extracellular)'}
ggplot(chr_expr_averages, aes(x=stage, y=total_expr, fill=stage)) + 
    geom_bar(stat='identity', color=CONFIG$stroke) +
    scale_fill_manual(values=CONFIG$colors) +
    theme(axis.text.x=element_text(angle=90)) +
    facet_wrap(~chr)
```

## Average RPKM expression by chromosome (developmental stage)

Below, a more granular look at chromosome-level average expression is shown for
each developmental stage studied.

```{r figure_X15_average_chr_gene_rpkm_expr_condition_specific, fig.cap='**Figure X15** Average RPKM expression by chromosome at the gene level (by developmental stage)'}
# Compute average total RPKM for each chromosome (developmental stage)
chr_expr_averages <- rpkm_counts_long %>% 
    group_by(chr, condition) %>%
    summarize(average_expr=median(expr),
              total_expr=sum(expr))

ggplot(chr_expr_averages, aes(x=condition, y=total_expr, fill=condition)) + 
    geom_bar(stat='identity', color=CONFIG$stroke) +
    theme(axis.text.x=element_text(angle=90)) +
    facet_wrap(~chr)
```

## Total RPKM expression by chromosome (developmental stage)

Similar to the above plot, but using _total_ expression instead of _average_
expression.

```{r figure_X16_total_chr_gene_rpkm_expr_condition_specific, fig.cap='**Figure X16** Total RPKM expression by chromosome at the gene level (by developmental stage)'}
ggplot(chr_expr_averages, aes(x=condition, y=total_expr, fill=condition)) + 
    geom_bar(stat='identity', color=CONFIG$stroke) +
    theme(axis.text.x=element_text(angle=90)) +
    facet_wrap(~chr)
```

## Average RPKM expression per nucleotide by PTU

Next, a table of putative polycistronic transcriptional units was creating by
grouping contiguous sets of genes all on the same strand (script and tables
available at https://github.com/elsayed-lab/ptu-detection), and average 
expression was plotted as a function of PTU.

This can be useful to detect trends in expression at the level of PTUs.

```{r figure_X17_average_rpkm_expr_per_nt_by_ptu, fig.cap='**Figure X17** Average RPKM expression by PTU at the nucleotide level'}
# Add PTU mapping to RPKM gene counts
rpkm_totals <- merge(rpkm_totals, ptu_mapping, by.x='gene', by.y='gene_id')

# Determine the average expression by PTU
ptu_expr_averages <- rpkm_totals %>% 
    group_by(ptu) %>%
    summarise(average_expr=median(expr),
              total_expr=sum(expr))

# Add chromosome information
ptu_expr_averages <- merge(ptu_expr_averages, 
                           rpkm_totals %>% select(chr, ptu) %>% unique, by='ptu')

# Compute the combined length of CDS's for each PTU
ptu_cds_total_lengths <- merge(gene_lengths, ptu_mapping, by.x='gene', by.y='gene_id') %>%
    group_by(ptu) %>%
    summarise(total_length=sum(length))

ptu_stats <- merge(ptu_cds_total_lengths, ptu_expr_averages, by='ptu') %>%
    mutate(expr_per_nt=total_expr/total_length)

# ptu display order
ptu_order <- sort(unique(ptu_stats$ptu))

ggplot(ptu_expr_averages, aes(x=ptu, y=average_expr)) + 
    geom_bar(aes(fill=chr), stat='identity') +
    scale_x_discrete(limits=ptu_order) +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank()) +
    xlab("PTU") +
    ylab("Median Gene Expression (RPKM)") +
    ggtitle("Average RPKM expression by PTU")
```

## Total RPKM expression per nucleotide by PTU

Same as above, but showing _total_ expression instead of _average_ expression.

```{r figure_X18_total_rpkm_expr_per_nt_by_ptu, fig.cap='**Figure X18** Total RPKM expression by PTU at the nucleotide level'}
#ptu_stats <- merge(ptu_stats, rpkm_totals %>% select(chr, ptu) %>% unique, by='ptu')

# in order to avoid noise due to small, possibly incorrect PTU's, we will 
# exclude any PTUs with less than 5 genes
ptu_num_genes <- (ptu_mapping  %>% group_by(ptu) %>% summarize(num_genes=n()))
ptu_stats <- merge(ptu_stats, ptu_num_genes, by='ptu') %>%
    filter(num_genes >= 5)

ggplot(ptu_stats, aes(x=ptu, y=expr_per_nt)) +
    geom_bar(aes(fill=chr), stat='identity') +
    scale_x_discrete(limits=ptu_order) +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank()) +
    xlab("polycistronic transcriptional unit (PTU)") +
    ylab("RPKM Expression (per nt)") +
    ggtitle("Average RPKM expression by PTU, scaled by PTU length")

```
