## Average gene structure

```{r figure_X1_tcruzi_gene_structure_overview, fig.height=250/96, fig.cap="**Figure X1**: Median 5'UTR, CDS, and 3â€² UTR lengths across all development stages."}
# Summary of gene structure
polypyrimidine_tracts <- polypyrimidine_tracts %>%
    mutate(length=end - start)
pp_tract_summary <- polypyrimidine_tracts %>%
    group_by(stage) %>% 
    summarise(median=median(length, na.rm=TRUE))

median_5utr <- (utr5_summary %>% filter(stage == 'combined'))$median
median_cds  <- median(as.numeric(genes$size))
median_3utr <- (utr3_summary %>% filter(stage == 'combined'))$median
median_ppt  <- (pp_tract_summary %>% filter(stage == 'combined'))$median

plot_gene_structure(median_5utr, median_cds, median_3utr,
                    utr_color=CONFIG$colors[2], cds_color=CONFIG$colors[1])
```

Above, the median values for each gene structure feature (CDS, 5' UTR, and 3' UTR)
are independently computed. In order to get a sense for what an "average" gene
looks like, below the median transcript length is computed for the set of genes
for which all three lengths are available.

```{r, results='asis', message=FALSE}
# create a data frame to store gene-matched lengths
matched_lengths <- data.frame()

# compute median mRNA length for all genes for which we have sufficient
# information
for (gene_id in utr3_stats_combined$gene) {
    # skip genes for which we don't know both UTR lengths
    if (!gene_id %in% utr5_stats_combined$gene) {
        next
    }

    # get length of each component and add entry to dataframe
    matched_lengths <- rbind(matched_lengths, c(
        (utr5_stats_combined %>% filter(gene == gene_id))$length,
        (utr3_stats_combined %>% filter(gene == gene_id))$length,
        width(genes[genes$ID == gene_id])
    ))
}

cat(sprintf('- Median mRNA length: %d nt\n',
            as.integer(median(rowSums(matched_lengths)))))

# average proportion of mRNA contributed by each portion
utr5_prop <- 100 * mean(matched_lengths[,1] / rowSums(matched_lengths))
utr3_prop <- 100 * mean(matched_lengths[,2] / rowSums(matched_lengths))
cds_prop  <- 100 * mean(matched_lengths[,3] / rowSums(matched_lengths))

cat(sprintf('    - 5\'UTR: %0.2f %%\n', utr5_prop))
cat(sprintf('    - 3\'UTR: %0.2f %%\n', utr3_prop))
cat(sprintf('    - CDS: %0.2f %%\n\n', cds_prop))
```

## Average GC- and CT-richness

```{r table_S5, results='asis'}
# function to compute proportion of a nucleotide for a collection of sequences
nt_proportion <- function (seqs) { 
    # Exclude any N's
    seqs <- gsub('N', '', seqs)

    # feature lengths
    seq_lengths <- unlist(lapply(seqs, nchar))

    # measure proportion of each nucleotide, for each feature
    sapply(c('A', 'C', 'G', 'T'), function (nt) { 
        str_count(seqs, nt) / seq_lengths 
    })
}

# intergenic region composition
intergenic_seqs <- (intercds_regions %>% filter(stage == 'combined'))$seq
intergenic_comp <- nt_proportion(intergenic_seqs)

# 5' and 3'UTR composition
utr5_comp <- nt_proportion(utr5_stats_combined$seq)
utr3_comp <- nt_proportion(utr3_stats_combined$seq)

# Drop any features for which there were only N's (rare)
utr5_comp <- utr5_comp[complete.cases(utr5_comp),]
utr3_comp <- utr3_comp[complete.cases(utr3_comp),]

# CDS composition
cds_comp <- nt_proportion(as.character(cds_fasta))

# Genome composition
genome_comp <- nt_proportion(as.character(fasta))

# data frame of sequence compositions, by feature
table_S5 <- data.frame(
    utr5=apply(utr5_comp, 2, median),
    cds=apply(cds_comp, 2, median),
    utr3=apply(utr3_comp, 2, median),
    intergenic=apply(intergenic_comp, 2, median),
    genome=apply(genome_comp, 2, median)
)

# GC richness
table_S5 <- rbind(table_S5, GC=table_S5[2,] + table_S5[3,])

xkable(table_S5, caption="**Table S5** Median GC- and CT-richness for gene features.")
```
