## Summary of UTR and CDS lengths

### All genes

Median CDS and UTR lengths, considering only UTRs with support from three
or more reads.

```{r table_01, results='asis'}
# Table 1: Gene Structure Summary
cds_lengths <- data.frame(
    gene=genes$ID,
    CDS=width(genes)
)

gene_structure_lengths <- merge(merge(utr5_stats_combined %>% select(gene, `5' UTR`=length), 
                                      utr3_stats_combined %>% select(gene, `3' UTR`=length), 
                                      by='gene', all=TRUE), cds_lengths, by='gene', all=TRUE)
table_01 <- gene_structure_summary(gene_structure_lengths %>% select(-gene))
xkable(table_01, caption='**Table 1** Gene structure summary table.')
```

### Three-way tritryp Orthologs

Below, we report the statistics when considering only those genes shared by 
*L. major*, *T. brucei*, and *T. cruzi*.

```{r table_s05, results='asis'}
table_s05 <- gene_structure_summary(gene_structure_lengths %>%
                                        filter(gene %in% tritryp_orthologs) %>%
                                        select(-gene))
xkable(table_s05, caption='**Table S5** Gene structure summary table (three-way orthologs).')
```

## Average lengths

```{r figure_01_average_lengths, fig.cap=sprintf('**Figure 1** Characterization of %s gene structure', CONFIG$parasite), warning=FALSE}
# Average 5′ UTR length
p1 <- ggplot(utr5_stats_combined, aes(x=length)) + 
        geom_histogram(aes(y=..ncount..), binwidth=20, fill=CONFIG$colors[1], colour=CONFIG$stroke, alpha=0.75) +
        xlim(0, 1500) +
        ggtitle('A') + theme(plot.title=element_text(hjust=0)) + 
        labs(x="5′ UTR length (nt)")

# Average 3′ UTR length
p2 <- ggplot(utr3_stats_combined, aes(x=length)) + 
        geom_histogram(aes(y=..ncount..), binwidth=20, fill=CONFIG$colors[1], colour=CONFIG$stroke, alpha=0.75) +
        xlim(0, 1500) +
        ggtitle('B') + theme(plot.title=element_text(hjust=0)) + 
        labs(x="3′ UTR length (nt)")

# Average Polypyrimidine tract length
p3 <- ggplot(polypyr_combined, aes(x=length)) + 
        geom_histogram(aes(y=..ncount..), binwidth=10, fill=CONFIG$colors[1], colour=CONFIG$stroke, alpha=0.75) +
        xlim(0, 300) +
        ggtitle('C') + theme(plot.title=element_text(hjust=0)) + 
        labs(x="Polypyrimidine tract length (nt)")

# Average CDS length
cds_lengths <- as.data.frame(as.numeric(genes$size))
colnames(cds_lengths) <- 'length'

p4 <- ggplot(cds_lengths, aes(x=length)) + 
        geom_histogram(aes(y=..ncount..), binwidth=20, fill=CONFIG$colors[1], colour=CONFIG$stroke, alpha=0.75) +
        xlim(0, 5000) +
        ggtitle('D') + theme(plot.title=element_text(hjust=0)) + 
        labs(x="CDS length (nt)")

# Average distance between primary SL site and polypyrimidine tract
p5 <- ggplot(polypyr_combined, aes(x=sl_dist)) +
        geom_histogram(aes(y=..ncount..), binwidth=10, fill=CONFIG$colors[1], colour=CONFIG$stroke, alpha=0.75) +
        xlim(0, 300) +
        ggtitle('E') + theme(plot.title=element_text(hjust=0)) + 
        labs(x="Distance between primary trans-splicing site and polypyrimidine tract (nt)")

# Average distance between primary poly(A) site and polypyrimidine tract
p6 <- ggplot(polypyr_combined, aes(x=polya_dist)) +
        geom_histogram(aes(y=..ncount..), binwidth=10, fill=CONFIG$colors[1], colour=CONFIG$stroke, alpha=0.75) +
        xlim(0, 1500) +
        ggtitle('F') + theme(plot.title=element_text(hjust=0)) + 
        labs(x="Distance between primary polyadenylation site and polypyrimidine tract (nt)")

# Average intergenic length (distance between SL and Poly(A) sites)
p7 <- ggplot(intercds_regions %>% filter(stage == 'combined'), 
             aes(x=intergenic_length)) +
        geom_histogram(aes(y=..ncount..), binwidth=10, fill=CONFIG$colors[1], colour=CONFIG$stroke, alpha=0.75) +
        xlim(0, 1500) +
        ggtitle('G') + theme(plot.title=element_text(hjust=0)) + 
        labs(x="Intergenic length (nt)")

grid.arrange(p1, p5, p2, p6, p3, p7, p4, ncol=2, left='Transcript density')

# normalize plot y-axis heights
#get_ylim <- function (plt) {
#    ggplot_build(plt)$panel$ranges[[1]]$y.range[2]
#}
#plt_ylim <- max(sapply(list(p1, p2, p3), get_ylim))

#grid.arrange(p1 + ylim(0, plt_ylim), 
#             p2 + ylim(0, plt_ylim),
#             p3 + ylim(0, plt_ylim), ncol=2)

#plt_ylim <- max(sapply(list(p4, p5, p6), get_ylim))

#grid.arrange(p4 + ylim(0, plt_ylim), 
#             p5 + ylim(0, plt_ylim), 
#             p6 + ylim(0, plt_ylim), ncol=2)
```

