## Summary of UTR and CDS lengths

### All genes

Median CDS and UTR lengths.

```{r table_X3, results='asis'}
# Table X3: Gene Structure Summary
cds_lengths <- data.frame(
    gene=genes$ID,
    CDS=width(genes)
)

gene_structure_lengths <- merge(merge(utr5_stats_combined %>% select(gene, `5' UTR`=length), 
                                      utr3_stats_combined %>% select(gene, `3' UTR`=length), 
                                      by='gene', all=TRUE), cds_lengths, by='gene', all=TRUE)
table_X3 <- gene_structure_summary(gene_structure_lengths %>% select(-gene))
xkable(table_X3, caption='**Table X3** CDS and UTR length summary statistics.')
```

#### SL sites

```{r table_S2}
# create table S2 and reorder so that combined sheet comes first
table_S2 <- sl_sites[c('combined', CONFIG$stages)]

i <- 1

for (stage in names(table_S2)) {
    # sort each sheet by gene id
    table_S2[[stage]] <- table_S2[[stage]] %>% arrange(gene)

    # show a preview of the table (full table is very long)
    caption <- sprintf('**Table S2%s**: Coordinates for all detected trans-splicing acceptor sites (%s: First 10 entries shown).', letters[i], stage)
    xkable(table_S2[[stage]], caption=caption)
    i <- i + 1
}
```

#### Poly(A) sites

```{r table_S3}
# create table S3 and reorder so that combined sheet comes first
table_S3 <- polya_sites[c('combined', CONFIG$stages)]

i <- 1

for (stage in names(table_S3)) {
    # sort each sheet by gene id
    table_S3[[stage]] <- table_S3[[stage]] %>% arrange(gene)

    # show a preview of the table (full table is very long)
    caption <- sprintf('**Table S3%s**: Coordinates for all detected polyadenylation sites (%s: First 10 entries shown).', letters[i], stage)
    xkable(table_S3[[stage]], caption=caption)
    i <- i + 1
}
```

### Three-way tritryp Orthologs

Below, we report the statistics when considering only those genes shared by 
*L. major*, *T. brucei*, and *T. cruzi*.

```{r table_S4, results='asis'}
table_S4 <- gene_structure_summary(gene_structure_lengths %>%
                                        filter(gene %in% tritryp_orthologs) %>%
                                        select(-gene))
xkable(table_S4, caption='**Table S4** CDS and UTR length summary statistics (only including tritryp three-way ortholog genes).')
```

## Average lengths

```{r figure_1_average_lengths, fig.cap=sprintf('**Figure 1** Characterization of %s gene structure', CONFIG$parasite), warning=FALSE}
# Plot margins
plt_margins <- unit(c(6, 12, 6, 12), "pt")
        
# Average 5′ UTR length
p1 <- ggplot(utr5_stats_combined, aes(x=length)) + 
        geom_histogram(aes(y=..ncount..), binwidth=15, fill=CONFIG$colors[1], colour=CONFIG$stroke, alpha=0.75) +
        scale_x_continuous(limits=c(0, 1500), expand = c(0, 0)) +
        theme(axis.title.y=element_blank(), plot.margin=plt_margins) + 
        ggtitle('A') + theme(plot.title=element_text(hjust=0)) + 
        labs(x="5′ UTR length (nt)")

# Average 3′ UTR length
p2 <- ggplot(utr3_stats_combined, aes(x=length)) + 
        geom_histogram(aes(y=..ncount..), binwidth=15, fill=CONFIG$colors[1], colour=CONFIG$stroke, alpha=0.75) +
        scale_x_continuous(limits=c(0, 1500), expand = c(0, 0)) +
        theme(axis.title.y=element_blank(), plot.margin=plt_margins) + 
        ggtitle('B') + theme(plot.title=element_text(hjust=0)) + 
        labs(x="3′ UTR length (nt)")

# Average CDS length
cds_lengths <- as.data.frame(as.numeric(genes$size))
colnames(cds_lengths) <- 'length'

p3 <- ggplot(cds_lengths, aes(x=length)) + 
        geom_histogram(aes(y=..ncount..), binwidth=50, fill=CONFIG$colors[1], colour=CONFIG$stroke, alpha=0.75) +
        scale_x_continuous(limits=c(0, 5000), expand=c(0, 0)) +
        theme(axis.title.y=element_blank(), plot.margin=plt_margins) + 
        ggtitle('C') + theme(plot.title=element_text(hjust=0)) + 
        labs(x="CDS length (nt)")

# Average Polypyrimidine tract length
p4 <- ggplot(polypyr_combined, aes(x=length)) + 
        geom_histogram(aes(y=..ncount..), binwidth=3, fill=CONFIG$colors[1], colour=CONFIG$stroke, alpha=0.75) +
        scale_x_continuous(limits=c(0, 300), expand = c(0, 0)) +
        theme(axis.title.y=element_blank(), plot.margin=plt_margins) + 
        ggtitle('D') + theme(plot.title=element_text(hjust=0)) + 
        labs(x="PolyPy tract length (nt)")


# Average distance between primary SL site and polypyrimidine tract
p5 <- ggplot(polypyr_combined, aes(x=sl_dist)) +
        geom_histogram(aes(y=..ncount..), binwidth=4, fill=CONFIG$colors[1], colour=CONFIG$stroke, alpha=0.75) +
        scale_x_continuous(limits=c(0, 400), expand = c(0, 0)) +
        theme(axis.title.y=element_blank(), plot.margin=plt_margins) + 
        ggtitle('E') + theme(plot.title=element_text(hjust=0)) + 
        labs(x="Distance between primary trans-splicing site and PolyPy tract (nt)")

# Average distance between primary poly(A) site and polypyrimidine tract
p6 <- ggplot(polypyr_combined, aes(x=polya_dist)) +
        geom_histogram(aes(y=..ncount..), binwidth=15, fill=CONFIG$colors[1], colour=CONFIG$stroke, alpha=0.75) +
        scale_x_continuous(limits=c(0, 1500), expand = c(0, 0)) +
        theme(axis.title.y=element_blank(), plot.margin=plt_margins) + 
        ggtitle('F') + theme(plot.title=element_text(hjust=0)) + 
        labs(x="Distance between primary Poly(A) addition site and PolyPy tract (nt)")

# Average intergenic length (distance between SL and Poly(A) sites)
p7 <- ggplot(intercds_regions %>% filter(stage == 'combined'), 
             aes(x=intergenic_length)) +
        geom_histogram(aes(y=..ncount..), binwidth=15, fill=CONFIG$colors[1], colour=CONFIG$stroke, alpha=0.75) +
        scale_x_continuous(limits=c(0, 1500), expand = c(0, 0)) +
        theme(axis.title.y=element_blank(), plot.margin=plt_margins) + 
        ggtitle('G') + theme(plot.title=element_text(hjust=0)) + 
        labs(x="Intergenic length (nt)")

grid.arrange(p1, p5, p2, p6, p3, p7, p4, ncol=2, left='Transcript density')

# normalize plot y-axis heights
#get_ylim <- function (plt) {
#    ggplot_build(plt)$panel$ranges[[1]]$y.range[2]
#}
#plt_ylim <- max(sapply(list(p1, p2, p3), get_ylim))

#grid.arrange(p1 + ylim(0, plt_ylim), 
#             p2 + ylim(0, plt_ylim),
#             p3 + ylim(0, plt_ylim), ncol=2)

#plt_ylim <- max(sapply(list(p4, p5, p6), get_ylim))

#grid.arrange(p4 + ylim(0, plt_ylim), 
#             p5 + ylim(0, plt_ylim), 
#             p6 + ylim(0, plt_ylim), ncol=2)
```

