## Data types

This analysis makes use of a number of different data sources, some of which
were produced during previous analysis steps.

The sources of data and annotations are:

1. [TriTrypDB genome sequence (FASTA)](http://tritrypdb.org/tritrypdb/)
2. [TriTrypDB genome annotations (GFF)](http://tritrypdb.org/tritrypdb/)
3. Stage-specific spliced leader and poly(A) site CSV files)
   - Separate files for each stage, and for each site type (SL/Poly(A).
   - One entry for each detected site, including read coverage information.
4. Summary 5'UTR, 3′ UTR, trans-splicing, polyadenylation, inter-CDS,
    and polypyrimidine tract statistics.
   - 5' and 3′ UTRs as defined by the primary acceptor and poly-adenylation
     sites detected when aggregating data across all parasite developmental
     stages.

To generate the above data files described in 3) and 4) above from the original
RNA-Seq samples, the following two analysis pipelines may be applied sequently:

1. [UTR analysis pipeline](https://github.com/elsayed-lab/utr_analysis)
   - Analyzes FASTQ files and detects trans-splicing and polyadenylation
     events.
   - Outputs GFF files containing all detected sites.
2. [Gene structure determination pipeline](https://github.com/elsayed-lab/gene-structure-determination)
   - Takes the raw trans-splicing site and polyadenylation site data from the
     UTR analysis pipeline and defines primary UTR boundaries for each gene
     where sufficient data is available.
   - Outputs include:
        - An extended version of a TriTrypDB GFF file containing the annotated
          UTR bounaries
        - CSV summary statistics for _primary_ 5' and 3′ UTR boundaries
          (`5utr_stats.csv` and `3utr_stats.csv`)
        - CSV files containing _all_ trans-splicing and polyadenylation sites
          for the inter-CDS regions where primary UTR boundaries could be
          assigned, with each site assigned as being 'primary' or
          'alternative'.
        - CSV file containing information about all detected polypyrimidine
          tracts
        - CSV file containing information about inter-CDS and intergenic
          regions.

## Load sequence and annotations

```{r load_annotations, warning=FALSE}
# Load genome sequence and annotations
gff <- import.gff3(CONFIG$input_gff)
fasta <- readDNAStringSet(CONFIG$input_fasta)

# individual CDS sequences
cds_fasta <- readDNAStringSet(CONFIG$input_cds_fasta)

# fix cds identifiers so they match gene ids
names(cds_fasta) <- str_replace(names(cds_fasta), '\\.1 ', ' ')
names(cds_fasta) <- str_replace(names(cds_fasta), '[: ].*', '')

# Sort GFF GenomicRanges
gff <- sortSeqlevels(gff)
gff <- sort(gff, ignore.strand=TRUE)

chromosomes <- gff[gff$type %in% c('contig', 'chromosome')]
genes       <- gff[gff$type == 'gene']

# Remove noncoding RNAs from annotations
gene_types <- AnnotationDbi::select(orgdb, keys=keys(orgdb, 'GID'), 
                                    keytype='GID', columns='TYPE')
protein_coding <- gene_types$GID[gene_types$TYPE == 'proteincoding']

genes <- genes[genes$ID %in% protein_coding]

# remaining gene ids
gene_ids <- genes$ID

# Fix TriTrypDB FASTA names
# e.g. "LmjF.24 | organism=Leishmania_major_strain_Friedlin |..." -> "LmjF.24"
names(fasta) <- sapply(strsplit(names(fasta), ' | '), function(x) {x[1]})

# For L. major GFF including unannotated ORFs, compute size
if (sum(is.na(genes$size)) > 0) {
    missing_size <- is.na(genes$size)
    genes$size[missing_size] <- end(genes[missing_size]) - start(genes[missing_size])
}

# L. major: load unannotated ORFs detected from ribosome profiling data
#if (CONFIG$parasite == 'L. major Friedlin') {
#    orfs      <- gff[gff$type == 'ORF']
#    orfs$Name <- orfs$Name
#    orfs$description <- 'Unannotated ORF'

#    # 2015/06/30 A few of the unannotated ORFs appear to have multiple conflicting
#    # entries -- removing these for now...
#    orfs <- orfs[!duplicated(orfs$Name)]

#    # Drop GFF columns not shared between TriTrypDB GFF and uORF GFF
#    keep_cols <- intersect(colnames(mcols(genes)), colnames(mcols(orfs)))
#    genes <- genes[,keep_cols]
#    orfs <- orfs[,keep_cols]

#    genes <- append(genes, orfs)

#    # Fix names (L. major chromosome identifiers)
#    names(fasta) <- substring(names(fasta), 0, 7)
#}
```

```{r}
# create a gene/chromosome mapping
chr_mapping <- data_frame(gene=genes$ID, chr=as.character(genes@seqnames))
```

## Create a dataframe containing strand information for each gene

```{r}
gene_strands <- data_frame(gene=genes$ID, strand=as.vector(genes@strand))
```

## Load spliced leader and poly(A) site coordinates

```{r load_acceptor_sites, warning=FALSE}
sl_sites <- list()
polya_sites <- list()

# names of developmental stages being analyzed
stages <- c(CONFIG$stages, 'combined')

for (stage in stages) {
    # load site information
    sl_sites[[stage]]    <- read_csv(CONFIG$sl_sites[[stage]])
    polya_sites[[stage]] <- read_csv(CONFIG$polya_sites[[stage]])

    # remove sites associated with noncoding RNAs (occurs quite rarely)
    sl_sites[[stage]] <- sl_sites[[stage]][sl_sites[[stage]]$gene %in% protein_coding,]
    polya_sites[[stage]] <- polya_sites[[stage]][polya_sites[[stage]]$gene %in% protein_coding,]
}
```

## Load polypyrimidine tract information

```{r load_polypyr_tracts, warning=FALSE}
polypyrimidine_tracts <- cbind(read_csv(CONFIG$polypyrimidine_tracts[['combined']]),
                               stage='combined')

polypyrimidine_tracts <- polypyrimidine_tracts[polypyrimidine_tracts$primary_sl %in% sl_sites[['combined']]$site,]

for (stage in CONFIG$stages) {
    dat <- read_csv(CONFIG$polypyrimidine_tracts[[stage]])

    # remove polypyr tracts associated with SL sites that have been filtered out
    dat <- dat[dat$primary_sl %in% sl_sites[[stage]]$site,]
    polypyrimidine_tracts <- rbind(polypyrimidine_tracts, cbind(dat, stage=stage))
}

# combined statistics for polypyrimidine tracts
polypyrimidine_tracts <- polypyrimidine_tracts %>% 
    mutate(length=end - start)

polypyr_combined <- polypyrimidine_tracts %>% 
    filter(stage == 'combined')
```

## Load inter-CDS and intergenic region information

```{r load_intercds_regions, warning=FALSE}
intercds_regions <- cbind(read_csv(CONFIG$intercds_regions[['combined']]),
                                   stage='combined')

# filter out regions associated with at least one noncoding gene
intercds_regions <- intercds_regions[intercds_regions$left_gene %in% protein_coding & 
                                     intercds_regions$right_gene %in% protein_coding,]

for (stage in CONFIG$stages) {
    dat <- read_csv(CONFIG$intercds_regions[[stage]])

    # filter out regions associated with at least one noncoding gene
    dat <- dat[dat$left_gene %in% protein_coding & dat$right_gene %in% protein_coding,]
    intercds_regions <- rbind(intercds_regions, cbind(dat, stage=stage))
}
```

### Load 5' and 3′ UTR summary statistics

Summary statistics describing the _primary_ 5' and 3′ UTR boundaries.

```{r}
# Create data.frames containing the primary 5' and 3′  length stats
utr5_stats <- cbind(read_csv(CONFIG$utr5_stats[['combined']]), stage='combined')

for (stage in CONFIG$stages) {
    utr5_stats <- rbind(utr5_stats, 
                        cbind(read_csv(CONFIG$utr5_stats[[stage]]), stage=stage))
}

utr3_stats <- cbind(read_csv(CONFIG$utr3_stats[['combined']]), stage='combined')
for (stage in CONFIG$stages) {
    utr3_stats <- rbind(utr3_stats,
                        cbind(read_csv(CONFIG$utr3_stats[[stage]]), stage=stage))
}

# Filter out UTR stats for noncoding genes (around 50 genes for T. cruzi)
utr5_stats <- utr5_stats[utr5_stats$gene %in% protein_coding,]
utr3_stats <- utr3_stats[utr3_stats$gene %in% protein_coding,]

# Remove any genes for which UTR boundaries were assigned within one or more
# of the developmental stages, but could not be assigned when data from all
# stages were considered together (e.g. primary sites with most support in 
# wrong orientation). This occurs infrequently (~1.5% of assignments for T. cruzi).
utr5_combined_genes <- (utr5_stats %>% filter(stage == 'combined'))$gene
utr5_stats <- utr5_stats %>%
    filter(gene %in% utr5_combined_genes)

utr3_combined_genes <- (utr3_stats %>% filter(stage == 'combined'))$gene
utr3_stats <- utr3_stats %>%
    filter(gene %in% utr3_combined_genes)

# Add NA placeholders for genes where the feature length was determined in 
# one stage, but not one or more of the others
all_utr5 <- unique(utr5_stats$gene)
all_utr3 <- unique(utr3_stats$gene)

for (stg in stages) {
    # grab stage-specific 5'UTR stats
    stage_specific_utr5 <- utr5_stats %>% 
        filter(stage == stg)
    stage_specific_utr3 <- utr3_stats %>% 
        filter(stage == stg)

    # create rows for each gene missing in the stage
    utr5_missing <- all_utr5[!all_utr5 %in% stage_specific_utr5$gene]

    if (length(utr5_missing) > 0) {
        placeholders <- cbind(utr5_missing, NA, NA, NA, NA, NA, NA, stg)
        colnames(placeholders) <- colnames(utr5_stats)
        utr5_stats <- rbind(utr5_stats, placeholders)
    }

    utr3_missing <- all_utr3[!all_utr3 %in% stage_specific_utr3$gene]

    if (length(utr3_missing) > 0) {
        placeholders <- cbind(utr3_missing, NA, NA, NA, NA, NA, NA, stg)
        colnames(placeholders) <- colnames(utr3_stats)

        # add to main dataframe
        utr3_stats <- rbind(utr3_stats, placeholders)
    }
}

# Fix types
for (x in c('length', 'num_reads', 'gc', 'ct')) {
    utr5_stats[[x]] <- as.numeric(utr5_stats[[x]])
    utr3_stats[[x]] <- as.numeric(utr3_stats[[x]])
}

# For convenience, we will also create version of the above containing just the
# combined stats
utr5_stats_combined <- utr5_stats %>% filter(stage == 'combined')
utr3_stats_combined <- utr3_stats %>% filter(stage == 'combined')
```

```{r filter_sites, eval=FALSE}
# Disabled Nov 28, 2016
# Because there has already been a significant amount of vetting in the earlier
# steps of this analysis, and in order to simplify the text and capture as much
# site information as possible, this filtering is no longer used.

### Filter out low confidence UTR assignments

#Create alternate versions of the 5' and 3′ UTR statistics data frames which only
#include sites with some minimum specified coverage level.

# Create a version of thee stats with low-support sites filtered out; in
# general we will use the unfiltered data frame when computing coverage
# statistics, and the filtered version when assigning high-confidence UTR
# boundaries, etc.
utr5_low_count_genes <- (utr5_stats %>% 
                            filter(stage=='combined' & 
                                   num_reads < CONFIG$min_read_support_5utr))$gene
utr5_stats_filtered <- utr5_stats %>% filter(!gene %in% utr5_low_count_genes)

utr3_low_count_genes <- (utr3_stats %>% 
                            filter(stage=='combined' & 
                                   num_reads < CONFIG$min_read_support_3utr))$gene
utr3_stats_filtered <- utr3_stats %>% filter(!gene %in% utr3_low_count_genes)

utr5_stats_combined_filtered <- utr5_stats_filtered %>% filter(stage == 'combined')
utr3_stats_combined_filtered <- utr3_stats_filtered %>% filter(stage == 'combined')
```

### Load RNA-Seq read counts and mapping statistics

```{r load_raw_rnaseq_counts}
count_table <- read.csv(CONFIG$count_table, row.names=1)
```

Total number of RNA-Seq reads used in analysis and total number of reads which
mapped to the host and parasite genomes.

```{r load_mapping_stats}
num_reads_host      <- read_csv(CONFIG$reads_mapped_host)
num_reads_parasite  <- read_csv(CONFIG$reads_mapped_parasite)
num_reads_total     <- read_csv(CONFIG$reads_mapped_total)
num_reads_sl        <- read_csv(CONFIG$reads_mapped_sl)
num_reads_polya     <- read_csv(CONFIG$reads_mapped_polya)

# create a combined version of the coverage stats data, including sample
# metadata
dfs <- list(CONFIG$samples, num_reads_host, num_reads_parasite,
            num_reads_total, num_reads_sl, num_reads_polya)
num_reads = Reduce(function(...) merge(..., all=T), dfs)

# add zero-count rows to host
parasite_only_samples <- num_reads_total$sample_id[!num_reads_total$sample_id 
                                                   %in% num_reads_host$sample_id]
num_reads_host <- rbind(num_reads_host, cbind(sample_id=parasite_only_samples, num_reads_host=0))
num_reads_host$num_reads_host <- as.integer(num_reads_host$num_reads_host)



# combine into a single data frame, including sample metadata
mapping_stats <- merge(merge(merge(merge(merge(CONFIG$samples, 
                                   num_reads_host, by='sample_id'),
                                   num_reads_parasite, by='sample_id'),
                                   num_reads_total, by='sample_id'),
                                   num_reads_sl, by='sample_id'),
                                   num_reads_polya, by='sample_id')

# num_parasite_reads initially only counts the non-SL/Poly(A) reads; update
# the totals to include these
mapping_stats$num_parasite_reads <- mapping_stats$num_reads_parasite +
                                    mapping_stats$num_reads_sl +
                                    mapping_stats$num_reads_polya

```

### Load polycistronic transcriptional unit (PTU) mapping

Next we will load a mapping of gene PTU's created in a separate analysis. The
`med_5` column name in the original file refers to the level of smoothing that
was used during PTU detection.

The mapping was

```{r}
ptu_mapping <- read_csv(CONFIG$ptu_mapping)
```

### Load TriTryp ortholog gene list

```{r}
tritryp_orthologs <- readLines('../input/shared/tritryp_orthologs.txt')
```

