### RNA-Seq coverage support for UTR boundaries

```{r coverage_mapped_read_densities, fig.show='hold', out.width='450px', fig.cap="**Figure 3**: Distribution of number of SL- and Poly(A)-containing reads."}
# number of reads containing SL- or Poly(A) sequence
hist(log(utr5_stats_combined$num_reads + 1), main='# SL-containing reads')
hist(log(utr3_stats_combined$num_reads + 1), main='# Poly(A)-containing reads')
```

```{r coverage_summary_table, results='asis'}
total_genes <- nrow(utr5_stats_combined)
nvals <- c(1, 2, 3, 4, 5, 10, 100, 1000) 

# % genes with N or more reads mapped to feature
sl_pcts <- c()
polya_pcts <- c()

for (nval in nvals) {
	sl_pcts <- append(sl_pcts, sum(utr5_stats_combined$num_reads >= nval) / total_genes)
}

for (nval in nvals) {
	polya_pcts <- append(polya_pcts, sum(utr3_stats_combined$num_reads >= nval) / total_genes)
}

sl_pcts    <- round(sl_pcts, 4) * 100
polya_pcts <- round(polya_pcts, 4) * 100

# summary data frame
df <- cbind(N=nvals, SL=sl_pcts, PolyA=polya_pcts)

kable(df, caption='**Table 4**: Percent genes containing N or more reads mapping to the specified feature.')
```
