```{r settings}
# Input files generated specifically for this analysis
input_dir <- file.path('..', 'input', 'tcruzi_ystrain')

# Input and output data used from previous analyses
utr_analysis_basedir <- file.path(Sys.getenv('RESEARCH'), '2016',
                                  '06-gene-structure/01-gene-structure-determination')
utr_analysis_suffix <- 'tcruzi_ystrain/sl-min5_polya-min5_mindiff-2'
utr_analysis_indir  <- file.path(utr_analysis_basedir, 'input', utr_analysis_suffix)
utr_analysis_outdir <- file.path(utr_analysis_basedir, 'output', utr_analysis_suffix)

# Sample metadata

CONFIG <- list(
    # species information
    'parasite' = 'T. cruzi Y strain',
    'host'     = 'H. sapiens',
    'stages'  = c('trypomastigote', 'amastigote', 'epimastigote'),

    # Discrete color palette and stroke color to use for plots

    # http://www.colourlovers.com/palette/443995/i_demand_a_pancake
    'colors' = c('#45ADA8','#9DE0AD','#547980','#E5FCC2'),
    'stroke' = '#594F4F',

    # Output verbosity
    'verbose' = FALSE,

    # Sample metadata
    'samples' = read.csv('tcruzi_ystrain_samples.csv'),

    # Order to use when displaying conditions
    'condition_order' = c('trypo', 'epi', '4', '6', '12', '20', '24', '48', '72'),

    # output file prefix
    'outfile_prefix' = 'tcruzi_ystrain',

    # TriTrypDB GFF genome annotations (all stages)
    'input_gff' = file.path(utr_analysis_outdir, 'combined', 'TriTrypDB-27_TcruziCLBrenerEsmeraldo-like-with-utrs.gff'),

    # TriTrypDB FASTA genome sequence
    'input_fasta' = file.path(Sys.getenv("REF"), 'tcruzi_clbrener_esmeraldo-like/genome/'),

    'utr5_stats' = list(
        amastigote    =file.path(utr_analysis_outdir, 'amastigote',     '5utr_stats.csv'),
        combined      =file.path(utr_analysis_outdir, 'combined',       '5utr_stats.csv'),
        trypomastigote=file.path(utr_analysis_outdir, 'trypomastigote', '5utr_stats.csv'),
        epimastigote  =file.path(utr_analysis_outdir, 'epimastigote',   '5utr_stats.csv')
    ),

    'utr3_stats' = list(
        amastigote    =file.path(utr_analysis_outdir, 'amastigote',     '3utr_stats.csv'),
        combined      =file.path(utr_analysis_outdir, 'combined',       '3utr_stats.csv'),
        trypomastigote=file.path(utr_analysis_outdir, 'trypomastigote', '3utr_stats.csv'),
        epimastigote  =file.path(utr_analysis_outdir, 'epimastigote',   '3utr_stats.csv')
    ),

    'sl_sites' = list(
        amastigote    =file.path(utr_analysis_indir, 'tcruzi_infecting_hsapiens_amastigote_sl_sorted.gff.gz'),
        trypomastigote=file.path(utr_analysis_indir, 'tcruzi_infecting_hsapiens_trypomastigote_sl_sorted.gff.gz'),
        epimastigote  =file.path(utr_analysis_indir, 'tcruzi_infecting_hsapiens_epimastigote_sl_sorted.gff.gz')
    ),

    'polya_sites' = list(
        amastigote    =file.path(utr_analysis_indir, 'tcruzi_infecting_hsapiens_amastigote_polya_sorted.gff.gz'),
        trypomastigote=file.path(utr_analysis_indir, 'tcruzi_infecting_hsapiens_trypomastigote_polya_sorted.gff.gz'),
        epimastigote  =file.path(utr_analysis_indir, 'tcruzi_infecting_hsapiens_epimastigote_polya_sorted.gff.gz')
    ),

    # Non-coding RNA filters
    'id_filter_string'   = 'rRNA|5SRRNA|snRNA|snoRNA|SLRNA|TRNA|SRP',
    'type_filter_string' = 'rRNAencoding|snRNAencoding|snoRNAencoding|tRNAencoding',

    # Minimum number of reads for UTR to be used for analysis
    'min_read_support_5utr'   = 3,
    'min_read_support_3utr'   = 3,

    # Number of reads mapped
    'reads_mapped_total'    = file.path(input_dir, 'S01_num_reads_total.csv'),
    'reads_mapped_host'     = file.path(input_dir, 'S01_num_reads_host.csv'),
    'reads_mapped_parasite' = file.path(input_dir, 'S01_num_reads_parasite.csv'),

    # Number of reads with Sl/Poly(A) sites detected
    'reads_mapped_sl'       = file.path(input_dir, 'S01_num_reads_sl.csv'),
    'reads_mapped_polya'    = file.path(input_dir, 'S01_num_reads_polya.csv')
)

# add intra/extracellular field to samples
CONFIG$samples$stage <- ifelse(CONFIG$samples$condition %in% c('epi', 'trypo'),
                               'Extracellular', 'Intracellular')
rm(utr_analysis_basedir, utr_analysis_indir, utr_analysis_outdir)
```

