```{r settings}
#
# v2.6
# ----
# - Table/figure numbering updates.
# 
# v2.5
# ----
# - Figure tweaks and table/figure numbering and caption updates.
#
# v2.4
# ----
# - Added T. cruzi sylvio comparison section
# - Switched to log scale for size/color for figure 3, hiding genes with site
#   switches that are 5nt or less apart.
# - Renamed table 2 -> table 1
#
# v2.3
# ----
# - Relabeling figures and tables; removing unused ones.
# - Figure 3 histograms modified so that only canonical vs. canonical and
#   non-canonical vs. non-canonical distances are displayed.
# - Primary to secondary (PtoS) ratio included in alt. splicing /
#   polyadenylation site plots.
#
# v2.2
# ----
# - Figure and table tweaks.
#
# v2.1
# ----
# - Few figure tweaks. 
#
# v2.0
# ----
# - Fixed couple bugs in upstream analysis steps (utr_analysis.py) which
# resulted in erronious and missed sites.
# - Relaxed minimum detection length criteria from 4 -> 3 nt long.
# - Switching to TriTrypDB version 29 annotations.
#
# v1.6
# ----
# - Generating high-quality (tiff) versions of all figures.
# - Modified site coordinate tables (S2/S3) to include chromosome information
#   and to exclude stage-specific site identifiers.
#
# v1.5
# ----
# - Disabled filtering of SL/Poly(A) sites with less than three reads prior
#   to computing statistics.
# - Removed old table S1 (coverage statistics); already covered in earlier
#   manuscript.
#
# v1.4
# ----
# - Fixed an issue with coordinates for 5'UTR acceptor site position.
#
# v1.3
# ----
# - Modified UTR boundary detection logic (gene-structure-determination) to
#   exclude SL and Poly(A) sites which would lead to UTRs containing >= 100
#   N's. This is one issue leading to unexpectedly long UTRs.
# - Increasing window-size when scanning for polypyrimidine tracts from 250 to
#   400 nt.
#

# Load orgdb (used to determine gene type)
library('org.TcCLB.esmer.tritryp.db')
orgdb <- get('org.TcCLB.esmer.tritryp.db')

# Analysis version
analysis_version <- '2.6'

# Input and output data used from previous analyses
utr_analysis_basedir <- file.path(Sys.getenv('RESEARCH'), '2016',
                                  '06-gene-structure/01-gene-structure-determination')
utr_analysis_suffix <- 'tcruzi_ystrain/sl-min3_polya-min3_mindiff-2'
utr_analysis_indir  <- file.path(utr_analysis_basedir, 'input', utr_analysis_suffix)
utr_analysis_outdir <- file.path(utr_analysis_basedir, 'output', utr_analysis_suffix)

# stage-specific inputs
stages = c('trypomastigote', 'amastigote', 'epimastigote', 'combined')

utr5_stats <- as.list(file.path(utr_analysis_outdir, stages, '5utr_stats.csv'))
names(utr5_stats) <- stages

utr3_stats <- as.list(file.path(utr_analysis_outdir, stages, '3utr_stats.csv'))
names(utr3_stats) <- stages

sl_sites <- as.list(file.path(utr_analysis_outdir, stages, 'trans_splice_sites.csv'))
names(sl_sites) <- stages

intercds_regions <- as.list(file.path(utr_analysis_outdir, stages, 'intercds_regions.csv'))
names(intercds_regions) <- stages

polya_sites <- as.list(file.path(utr_analysis_outdir, stages, 'polyadenylation_sites.csv'))
names(polya_sites) <- stages

polypyr_tracts <- as.list(file.path(utr_analysis_outdir, stages, 'polypyrimidine_tracts.csv'))
names(polypyr_tracts) <- stages

# Directory containing input files generated specifically for this analysis
input_dir <- file.path('..', 'input', 'tcruzi_ystrain')

# Sample metadata
CONFIG <- list(
    # species information
    'parasite' = 'T. cruzi Y strain',
    'host'     = 'H. sapiens',
    'stages'  = c('trypomastigote', 'amastigote', 'epimastigote'),

    # Discrete color palette and stroke color to use for plots

    # http://www.colourlovers.com/palette/443995/i_demand_a_pancake
    'colors' = c('#45ADA8','#9DE0AD','#547980','#E5FCC2'),
    'stroke' = '#594F4F',

    # Output verbosity
    'verbose' = FALSE,

    # Sample metadata
    'samples' = read.csv('tcruzi_ystrain_samples.csv'),

    # Order to use when displaying conditions
    'condition_order' = c('trypo', 'epi', '4', '6', '12', '24', '48', '72'),

    # human-readable condition names
    condition_mapping = data.frame(from=c('4', '6', '12', '20', '24', '48', '72',
                                          'epi', 'trypo'),
                                   to=c('Amastigote 04 hpi', 'Amastigote 06 hpi', 
                                        'Amastigote 12 hpi', 'Amastigote 20 hpi',
                                        'Amastigote 24 hpi', 'Amastigote 48 hpi',
                                        'Amastigote 72 hpi', 'Epimastigote',
                                        'Trypomastigote')),

    # output directory
    'cache_dir'  = 'README_cache/tcruzi',
    'output_dir' = file.path('output/tcruzi_ystrain', analysis_version),

    # TriTrypDB FASTA genome sequence
    'input_fasta' = file.path(Sys.getenv("REF"), 'tcruzi_clbrener_esmeraldo-like', 
                              'genome/TriTrypDB-29_TcruziCLBrenerEsmeraldo-like_Genome.fasta'),
    'input_cds_fasta' = file.path(Sys.getenv("REF"), 'tcruzi_clbrener_esmeraldo-like', 
                                  'genome/TriTrypDB-29_TcruziCLBrenerEsmeraldo-like_AnnotatedCDSs.fasta'),

    # Stage-specific GFFs containing genes and annotated UTRs
    'input_gff' = file.path(utr_analysis_outdir, 'combined',
                            'TriTrypDB-29_TcruziCLBrenerEsmeraldo-like-with-utrs.gff'),

    # Stage-specific GFFs containing SL and Poly(A) site annotations
    # only included here to be copied to the output dir for easier submission
    'site_gffs' = Sys.glob(file.path(utr_analysis_indir, '*.gff.gz')),

    # RNA-Seq count table
    'count_table' = file.path(Sys.getenv("SCRATCH"), 'tcruzi-hsapiens',
                              'counts', 'tcruzi-all-samples.csv'),

    # Stage-specific _primary_ SL/Poly(A) site and UTR statistics
    'utr5_stats' = utr5_stats,
    'utr3_stats' = utr3_stats,

    # Polycistronic transcriptional unit (PTU) mapping
    'ptu_mapping' = file.path(Sys.getenv('RESEARCH'), '2013', '09-polycistron_detection',
                              'mappings',
                              'TriTrypDB-29_TcruziCLBrenerEsmeraldo-like_PTUs.csv'),

    # Chromosome order to use for plotting
    'chr_order' = sprintf("TcChr%d-S", 1:41),

    # SL / Poly(A) detection data
    'sl_sites' = sl_sites,
    'polya_sites' = polya_sites,

    # Inter-CDS and intergenic region data
    'intercds_regions' = intercds_regions,

    # Polypyrimidine tract information
    'polypyrimidine_tracts' = polypyr_tracts,

    # Non-coding RNA filters
    'id_filter_string'   = 'rRNA|5SRRNA|snRNA|snoRNA|SLRNA|TRNA|SRP',
    'type_filter_string' = 'rRNAencoding|snRNAencoding|snoRNAencoding|tRNAencoding',

    # Minimum number of reads for UTR to be used for analysis
    'min_read_support_5utr'   = 5,
    'min_read_support_3utr'   = 5,

    # Number of reads mapped
    'reads_mapped_total'    = file.path(input_dir, 'S01_num_reads_total.csv'),
    'reads_mapped_host'     = file.path(input_dir, 'S01_num_reads_host.csv'),
    'reads_mapped_parasite' = file.path(input_dir, 'S01_num_reads_parasite.csv'),

    # Number of reads with Sl/Poly(A) sites detected
    'reads_mapped_sl'       = file.path(input_dir, 'S01_num_reads_sl.csv'),
    'reads_mapped_polya'    = file.path(input_dir, 'S01_num_reads_polya.csv')
)

# add intra/extracellular field to samples
CONFIG$samples$stage <- ifelse(CONFIG$samples$condition %in% c('epi', 'trypo'),
                               'Extracellular', 'Intracellular')

# create output directory if it doesn't already exist
if (!dir.exists(CONFIG$output_dir)) {
    dir.create(CONFIG$output_dir, recursive=TRUE)
}

# clean-up
rm(analysis_version, utr_analysis_basedir, utr_analysis_indir,
   utr_analysis_outdir, utr5_stats, utr3_stats, sl_sites, polya_sites,
   polypyr_tracts)
```

